File: java/min_stack.java
public class MinStack {

    ArrayDeque<Integer> stack;
    ArrayDeque<Integer> min;
    /** initialize your data structure here. */
    public MinStack() {
        stack=new ArrayDeque<>();
        min=new ArrayDeque<>();
    }
    
    public void push(int x) {
        stack.add(x);
        if (min.isEmpty()||x<=min.peekLast()) min.add(x);
    }
    
    public void pop() {
        if (!stack.isEmpty()) {
            if (min.peekLast().equals(stack.peekLast())) min.pollLast();
            stack.pollLast();
        }
    }
    
    public int top() {
        if (!stack.isEmpty()) {
            return stack.peekLast();
        } else {
            return -1;
        }
    }
    
    public int getMin() {
        if (!min.isEmpty()) {
            return min.peekLast();
        } else {
            return -1;
        }
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: python/min-stack.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.data, self.min=[], []

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        self.data.append(x)
        if len(self.min)==0 or x<=self.min[-1]: self.min.append(x)

    def pop(self):
        """
        :rtype: void
        """
        if len(self.data)==0: return
        if self.min[-1]==self.data[-1]: self.min.pop()
        self.data.pop()

    def top(self):
        """
        :rtype: int
        """
        if len(self.data)==0: return -1
        return self.data[-1]

    def getMin(self):
        """
        :rtype: int
        """
        if len(self.min)==0: return -1
        return self.min[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
